python/ray/tune/impl/tuner_internal.py

```diff
@@ -0,0 +1,178 @@
+import copy
+import os
+from typing import Any, Callable, Dict, Optional, Type, Union
+
+import ray.cloudpickle as pickle
+from ray.ml.config import RunConfig
+from ray.ml.trainer import Trainer
+from ray.tune import Experiment, TuneError
+from ray.tune.impl.utils import execute_dataset
+from ray.tune.result_grid import ResultGrid
+from ray.tune.trainable import Trainable
+from ray.tune.tune import run
+from ray.tune.tune_config import TuneConfig
+
+
+_TRAINABLE_PKL = "trainable.pkl"
+_TUNER_PKL = "tuner.pkl"
+_TRAINABLE_KEY = "_trainable"
+_PARAM_SPACE_KEY = "_param_space"
+
+
+class TunerInternal:
+    """The real implementation behind external facing ``Tuner``.
+
+    The external facing ``Tuner`` multiplexes between local Tuner and remote Tuner
+    depending on whether in Ray client mode.
+
+    In Ray client mode, external Tuner wraps ``TunerInternal`` into a remote actor,
+    which is guaranteed to be placed on head node.
+
+    TunerInternal can be constructed from fresh, in which case, `trainable` needs to
+    be provided, together with optional `param_space`, `tune_config` and `run_config`.
+
+    It can also be restored from a previous failed run (given `restore_path`).
```

Comment:
```suggestion
    TunerInternal can be constructed from fresh, in which case, ``trainable`` needs to
    be provided, together with optional ``param_space``, ``tune_config`` and ``run_config``.

    It can also be restored from a previous failed run (given ``restore_path``).
```